pandas/core/indexing.py

```diff
@@ -88,8 +89,8 @@ class IndexingError(Exception):
 
 
 class _NDFrameIndexer(_NDFrameIndexerBase):
-    _valid_types = None
-    _exception = KeyError
+    _valid_types = None  # type: Optional[str]
+    _exception = KeyError  # type: Type[Exception]
```

Comment:
Does inference not work here?

Reply:
Mypy infers the type as `KeyError`, but that's too narrow for the subclasses as written.

`_NDFrameIndexer` defines `_exception` as `KeyError`. It's subclass and sub-subclass each override their own `_exception`s differently. E.g., `_LocationIndexer(_NDFrameIndexer)` wants a plain `Exception`, and `_iLocIndexer(_LocationIndexer)` wants an `IndexError`.

For mypy to accept all three assignments to the same variable name in that inheritance hierarchy, it needs to be typed more broadly than `KeyError`. I chose `Exception` because that covers all the existing options.

Reply:
Thanks for clarifying. Are there any subclasses that would actually use this? Wondering if we should just change this to Exception rather than typing it as such; just seems kind of strange for a base class like his to have a specific Exception type that subclasses would loosen

Reply:
I agree it's odd that subclasses' errors get less specific, so I suspect it's just a design oversight or a function of the bits being written at different times. Loosening the ancestor is probably fine.